webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/reports/ComplianceLevel.scala

```diff
@@ -69,7 +74,114 @@ final case class CompliancePercent(
+   *   the last one will be false by 0.13%, which is ok)
+   * - the biggest value is rounded to compensate
+   * - in case of equality (ex: 1/3 success, 1/3 NA, 1/3 error), the biggest is
+   *   chosen accordingly to "getWorsteType" logic.
+   */
+  def fromLevels(c: ComplianceLevel): CompliancePercent = {
+    val MINPC = BigDecimal(0.01) // 0.01% since we keep two digits in percents
+
+    if(c.total == 0) {  // special case: let it be 0
+      CompliancePercent()
+    } else {
+      val total = c.total // not zero
+      def pc_for(i:Int) : Double = {
+        if(i == 0) {
+          0
+        } else {
+          val pc = (i * 100 / BigDecimal(total)).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          (if(pc < MINPC) MINPC else pc).toDouble
+        }
+      }
+
+      // default percent calculation
+      val levels = CompliancePercent.sortLevels(c)
+      // round the first ones keeping at least MINPC
+      val pc = levels.init.map { case (l, index) => (pc_for(l), index) }
+      // the last one is rounded by "100 - sum of other"
+      val pc_last = 100 - pc.map(_._1).sum
+
+      // check that the rounding error is below what is expected (0.15%) else warn
+      val error = Math.abs((pc_for(levels.last._1)-pc_last)/pc_last)
+      if( error > (levels.size+1)*MINPC) {
+        ComplianceLogger.info(s"Rounding error in compliances is above expected threshold: error is ${error} for ${c}")
+      }
+
+      // sort back percents by index
+      val all = ((pc_last, levels.last._2) :: pc).sortBy(_._2).map(_._1)
+
+      // create array of pecents
+      CompliancePercent.fromSeq(all)
+    }
+  }
+
+  /*
+   * returned a list of (level, index) where index correspond to the index of the
+   * level in the compliance array. The list is sorted in the relevant order so that
+   * the smaller level are at head, and in case of equality, the "worse order" order is
+   * kept (worse order last)
+   */
+  def sortLevels(c: ComplianceLevel): List[(Int, Int)] = {
+    // we want to compare accordingly to `ReportType.getWorsteType` but I don't see any
+    // way to do it directly since we don't use the same order in compliance.
+    // So we map index of a compliance element to it's worse type order and compare by index
+    val WORSE_ORDER = Seq(13, 4, 3, 12, 11, 2, 5, 6, 8, 0, 1, 9, 10, 7).zipWithIndex.toMap
+
+    val levels = List(
+        c.pending
+      , c.success
+      , c.repaired
+      , c.error
+      , c.unexpected
+      , c.missing
+      , c.noAnswer
+      , c.notApplicable
+      , c.reportsDisabled
+      , c.compliant
+      , c.auditNotApplicable
+      , c.nonCompliant
+      , c.auditError
+      , c.badPolicyMode
+    ).zipWithIndex
+
+    // sort smallest first, and then by worst type
+    levels.sortWith { case ((l1, i1), (l2, i2)) =>
+      if(l1 == l2)  {
+        WORSE_ORDER.getOrElse(i1, WORSE_ORDER.size) > WORSE_ORDER.getOrElse(i2, WORSE_ORDER.size)
```

Comment:
If levels were created with compliance levels from worse to best, and then zipWithIndexed, you could simply use
i1 > i2


this methods would be much easier, and skip the whole map key search

Reply:
I'm not sure I understand. Do you propose to change ComplianceLevel properties order? If not, can you explain again? If yes, I thought to that, but that order is used as an hypothesis for a lot of things, from test to backend to js, and I fear consequence would be far reaching. Plus, extension would mean changing relative order each time, so if we want to do that, we need to make the change index independent (ie forbids things like `ComplianceLevel(0,1,0)`, and it's  implicated with case classes)

Reply:
no, not changing the order of properties, but the order in the list of levels, to have the live easily ordered. 
But now I understand that there is no translation from that to the levels in the case class, so it would only means moving it away.. which may or may not be better

Reply:
OK, it was just displacing the problem. I did my other idea, and it works well.